### PR TITLE
arch: arm: define and implement ARM IRQ target state API

### PR DESCRIPTION
**Pull request by ioannisg**
This commit contributes the definition and the implementation
of an API for ARM MCUs to configure and get the configuration
of the security state targeted by a given external IRQ line.
The API is only available in Secure ARM Firmware.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Addresses #6730, please take a look into the details.

### CODE REVIEW

File: arch/arm/core/irq_manage.c [hunk 1]
```diff
@@ -188,3 +188,65 @@ void _arch_isr_direct_header(void)
...
+ * It shall assert if the operation is not performed successfully.
+ *
+ * @param irq IRQ line
+ * @param secure_state 1 if target state is Secure, 0 otherwise.
+ *
+ * @return N/A
+ */
+void irq_target_state_set(unsigned int irq, int secure_state)
+{
+	if (NVIC_GetTargetState(irq) == 0) {
```
Comment (pizi-nordic) [inline arch/arm/core/irq_manage.c]: Why we have to check current state of the IRQ?
Is there any HW limitation that prevents us to set the same state twice?

File: arch/arm/core/irq_manage.c [hunk 2]
```diff
@@ -188,3 +188,65 @@ void _arch_isr_direct_header(void)
...
+ * It is only compiled if ARM_SECURE_FIRMWARE is defined.
+ * It should only be called while in Secure state, otherwise, a read attempt
+ * to NVIC.ITNS register is read-as-zero(RAZ), as the ITNS register is not
+ * banked between security states and, therefore, has no Non-Secure instance.
+ *
+ * @param irq IRQ line
+ *
+ * @return 1 if target state is Non-Secure, 0 otherwise.
+ */
+int irq_target_state_is_nonsecure(unsigned int irq)
```
Comment (pizi-nordic) [inline arch/arm/core/irq_manage.c]: Are you planning irq_target_state_is_secure() ?
IMHO irq_target_state_is_secure() will better match irq_target_state_set() which uses true (non-zero) value to indicate secure IRQ.

File: arch/arm/core/irq_manage.c (hunk 1 above)
Comment (ioannisg) [reply arch/arm/core/irq_manage.c]: Not that I know of. Then, I can simplify this code, I suppose, removing the check on the current state.

File: arch/arm/core/irq_manage.c (hunk 2 above)
Comment (ioannisg) [reply arch/arm/core/irq_manage.c]: Right. I will keep this function as is, and, instead, align the above one to "match".

**Comment by ioannisg** [comment]
Re-based and addressed comments by @pizi-nordic

File: arch/arm/core/irq_manage.c (hunk 1 above)
Comment (pizi-nordic) [reply arch/arm/core/irq_manage.c]: IMHO we should use simpler approach.

File: arch/arm/core/irq_manage.c (hunk 2 above)
Comment (pizi-nordic) [reply arch/arm/core/irq_manage.c]: We can just provide 2 functions: _is_secure() and _is_nonsecure(). If we make _is_secure() inline there will be no penalty.

Comment (agross-linaro) [reply arch/arm/core/irq_manage.c]: i could go either way on two functions.  The plus side is that it's extremely clear.  But I don't have strong feelings about it.

**Review by agross-oss — APPROVED** [review]

File: arch/arm/core/irq_manage.c (hunk 2 above)
Comment (pizi-nordic) [reply arch/arm/core/irq_manage.c]: For me a `if (!irq_target_state_is_nonsecure(MY_IRQ))` is just awkward. A `if (irq_target_state_is_secure(MY_IRQ))` is much easier to read.

Comment (ioannisg) [reply arch/arm/core/irq_manage.c]: Then I will go with your suggestion and rename the functions, last argument makes sense, indeed.

**Review by pizi-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**